text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1722 @@
+```rust
+pub mod crate_a {
+
+    #[repr(C)]
+    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);
+
+    impl<'a, T> NonEmptySlice<'a, T> {
+        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
+            assert!(len <= N && len > 0);
+            Self(
+                private::NonEmptySliceInner {
+                    data: arr as *const T,
+                    len,
+                    lifetime: core::marker::PhantomData,
+                }
+            )
+        }
+
+        pub fn first(&self) -> &'a T {
+            unsafe { &*self.0.data }
+        }
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        #[repr(C)]
+        pub struct NonEmptySliceInner<'a, T> {
+            pub data: *const T,
+            pub len: usize,
+            pub lifetime: core::marker::PhantomData<&'a ()>,
+        }
+    }
+
+}
+```
+In the above example, the definitions of both `NonEmptySlice` and its field `NonEmptySliceInner` are marked `pub`, and all fields of these types are marked `pub`. However, `NonEmptySlice` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `NonEmptySliceInner` is not visibile outside of `crate_a`.
+
+#### Constructability and Transmutation
+Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author. In the previous examples, it would be *unsafe* to transmute `[usize; 2]` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would violate memory safety:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+// [usize; 2] ⟶ NonEmptySlice
+let _: NonEmptySlice<'static, u8> = transmute!([0usize; 2]); // Compile Error: `NonEmptySlice<_, _>` is not constructible here.
+```
+If a field is private, then instantiating or modifying it via transmutation is not, generally speaking, safe.
+
+For transmutations where the destination type involves mutate-able references, the constructability of the *source* type is also relevant. Consider:
+```rust
+/* ⚠️ This example intentionally does not compile. */
+let arr = [0u8, 1u8, 2u8];
+let mut x = NonEmptySlice::from_array(&arr, 2);
+{
+    // &mut NonEmptySlice ⟶ &mut [usize; 2]
+    let y : &mut u128 = transmute!(&mut x) // Compile Error! `&mut NonEmptySlice` is not constructible here.
+    *y[0] = 0;
+    *y[1] = 0;
+}
+
+let z : NonEmptySlice<u8> = x;
+```
+If this example did not produce a compile error, the value of `z` would not be a safe instance of its type, `NonEmptySlice`, because `z.first()` would dereference a null pointer.
+
+#### Constructability and Scope
```

Comment:
The @rust-lang/lang team [wondered](https://github.com/rust-lang/lang-team/issues/21#issuecomment-707301912):

> * One question that we were unsure about was whether `Here!` was meant to guarantee:
>   
>   * permitting transmute within a crate even if semver stability is not guaranteed (because you are in same compilation unit)
>   * or privacy (i.e., permitting transmute iff you have private access to the fields of a type)
>   * or both (or some other reasons)
>   * but perhaps we also just need to re-read RFC

The answer is **Just privacy**! Rust already has a notion of "you can fiddle with the fields arbitrarily", and that's visibility. Specifically, if you could already construct the type using its implicit constructor, then transmutation doesn't really gain you anything new. However, Rust's trait system isn't natively aware of constructability, because it isn't aware of scope. [This section of the RFC poses the problem.](https://github.com/rust-lang/rfcs/blob/bb9b988c10a60d714ad52c9804aad36961960736/text/0000-safer-transmute.md#constructability-and-scope) The extra `Scope` parameter and `Here()` enables `TransmuteFrom` to be scope-aware *without* significantly rewriting the trait system.

I'm hesitant to propose anything that might involve far-reaching changes to the trait system, but there *is* an alternative: make scope implicit (rather than explicit) by tracing the provenance of a `TransmuteFrom` bound and checking constructability at its source. This would be rather weird: `Foo: TransmuteFrom<Bar>` might be satisfied in one scope, but not in another! I don't know the details of how Rust's trait system works, but it would have to be *very* careful to never cache a proof that `TransmuteFrom<Src>` holds for a type — doing so would be unsound.